src/shared/utils/sqlInterval.ts

```diff
@@ -0,0 +1,88 @@
+import {TimeRange} from 'src/types'
+import {DEFAULT_TIME_RANGE} from 'src/shared/constants/timeRanges'
+
+const durationExpr = /^\-(([0-9]+)(y|mo|w|d|h|ms|s|m|us|µs|ns))+$/g
+const singleDuration = /(([0-9]+)(y|mo|w|d|h|ms|s|m|us|µs|ns))/g
+const nanosecondDuration = /(([0-9]+)(ns))/g
+const microsecondDuration = /(([0-9,\.]+)(\W?)(microsecond))/g
+
+const convertToInterval = (upperOrLower: string) => {
+  if (upperOrLower == null || upperOrLower == 'now()') {
+    return 'now()'
+  }
+  if (!!upperOrLower.match(durationExpr)) {
+    let interval = [
+      ...upperOrLower
+        .replace('-', '') // -12d6h => 12d6h
+        .matchAll(singleDuration), // [['12d','12','h'],['6h','6','h']]
+    ]
+      .map(x => x[0]) // because want most greedy match for each
+      .join(' ') // '12d 6h'
+      .replace('y', ' year')
+      .replace('h', ' hour') // match h, before month with h
+      .replace('mo', ' month')
+      .replace('w', ' week')
+      .replace('d', ' day')
+      .replace('ms', ' millisecond') // match ms, before m|s
+      .replace('us', ' microsecond')
+      .replace('µs', ' microsecond')
+      .replace('m ', ' minute ') // don't match `m` in millisecond, etc
+      .replace('s ', ' second ')
+
+    // postgres does not support nanoseconds. Convert.
+    let toReplace = interval.match(nanosecondDuration)
+    if (!!toReplace || toReplace?.length > 0) {
+      try {
+        toReplace.forEach(init => {
+          const replacement = `${
+            parseFloat(init.split('ns')[0]) / 1000
+          } microsecond`
+          interval = interval.replace(init, replacement)
+        })
+      } catch (e) {
+        throw new Error(`Cannot handle nanosecond: ${e}`)
+      }
+    }
+
+    // our flight sql is not supporting microseconds. Convert.
+    toReplace = interval.match(microsecondDuration)
+    if (!!toReplace || toReplace?.length > 0) {
+      try {
+        toReplace.forEach(init => {
+          const replacement = `${
+            parseFloat(init.split('microsecond')[0]) / 1000
+          } millisecond`
+          interval = interval.replace(init, replacement)
+        })
+      } catch (e) {
+        throw new Error(`Cannot handle microsecond: ${e}`)
+      }
+    }
+
+    return `now() - interval '${interval}'`
+  }
+
+  const timestamp = new Date(upperOrLower)
+  if (timestamp.toTimeString() === 'Invalid Date') {
+    throw new Error(`Unknown custom time: ${upperOrLower}`)
+  }
+  // TODO -- handle timezones.
+  return `timestamp '${timestamp.toISOString()}'`
+}
+
+/** Convert duration to postgres interval.
+ *     https://www.postgresql.org/docs/15/datatype-datetime.html#DATATYPE-INTERVAL-INPUT
+ */
+export const rangeToInterval = (range: TimeRange) => {
+  switch (range.type) {
+    case 'selectable-duration':
+      return range.sql
+    case 'custom':
+    case 'duration':
+      const upper = convertToInterval(range.lower)
+      const lower = convertToInterval(range.upper)
```

Comment:
Not sure if we want to use the try/catch, since it requires us to not error. Which translate to returning a default behavior (such as the default timeRange). I feel like instead we want the user to know that they entered in an invalid time. What do you think?

Reply:
Oh, I was asking the case what if `convertToInterval()` throw an error like this https://github.com/influxdata/ui/blob/6c01e6ed975a5aadc9dfe41e8d867630e7abe0d8/src/shared/utils/sqlInterval.ts#L44
should it be handled under `case 'duration':`? Something like:

```
case 'duration':
  try {
      const upper = convertToInterval(range.lower)
      const lower = convertToInterval(range.upper)
      return `time >= ${upper} AND time <= ${lower}`
  } catch (e) {
      ....
      return DEFAULT_TIME_RANGE.sql
  }
```

Reply:
Exactly.
But if return it as the default tiemRange --> we are silently doing things, and not telling the user why. So the user case to notice (on their own) that the inserted timeRange does not match.

Alternatively, we could check to see if this throws at a higher try/catch level, where we can toast to the user. Such as within the SqlConnector. We would then also want to reset the persistence context session back to the default timeRange -- but this persistence context only exists for the QxBuilder. And our connectors are right now designed to work inside this persistence context (e.g. the QxBuilder), as well as outside the persistence context (e.g. the other raw editors based on timeMachine use the Flux LspConnector for LSP actions outside of composition).

So that's where I'm not sure whether or not it's worthwhile to enter more code complexity. Especially since the `session.range` can only come from the datePicker, which already does not let an invalid duration (or date) be entered. This means that if we hit the current/existing throw clause -- it must be an error in our code.

Reply:
For the specific throwing for nanoseconds -- our code already converts to millisecond. So that throw clause is hit if (and only if) we have a bug in our replacement code.

Reply:
Ah, I got what you mean by "since it requires us to not error"! Yeah, it makes sense. Let's just keep it as is